storage/grpc_metrics.go

```diff
@@ -0,0 +1,161 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"strings"
+
+	mexporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"google.golang.org/api/option"
+	estats "google.golang.org/grpc/experimental/stats"
+	"google.golang.org/grpc/stats/opentelemetry"
+)
+
+const (
+	monitored_resource_name         = "storage.googleapis.com/Client"
+	mr_location_label_default       = "global"
+	mr_cloud_platform_label_default = "unknown"
+	mr_host_id_label_default        = "unknown"
+	metric_format_prefix            = "storage.googleapis.com/client/"
+	metric_lb_locality_label        = "grpc.lb.locality"
+)
+
+func getMetricsToEnable() []estats.Metric {
+	return []estats.Metric{"grpc.lb.wrr.rr_fallback",
+		"grpc.lb.wrr.endpoint_weight_not_yet_usable",
+		"grpc.lb.wrr.endpoint_weight_stale",
+		"grpc.lb.wrr.endpoint_weights",
+		"grpc.lb.rls.cache_entries",
+		"grpc.lb.rls.cache_size",
+		"grpc.lb.rls.default_target_picks",
+		"grpc.lb.rls.target_picks",
+		"grpc.lb.rls.failed_picks",
+		"grpc.xds_client.connected",
+		"grpc.xds_client.server_failure",
+		"grpc.xds_client.resource_updates_valid",
+		"grpc.xds_client.resource_updates_invalid",
+		"grpc.xds_client.resources",
+	}
+}
+
+func getMetricsEnabledByDefault() []estats.Metric {
+	return []estats.Metric{
+		"grpc.client.attempt.sent_total_compressed_message_size",
+		"grpc.client.attempt.rcvd_total_compressed_message_size",
+		"grpc.client.attempt.started",
+		"grpc.client.attempt.duration",
+		"grpc.client.call.duration",
+	}
+}
+
+func getViewMasks(metrics []estats.Metric) []sdkmetric.View {
+	views := []sdkmetric.View{}
+	for _, m := range metrics {
+		views = append(views, sdkmetric.NewView(sdkmetric.Instrument{
+			Name: string(m),
+		}, sdkmetric.Stream{Name: strings.ReplaceAll(string(m), ".", "/")}))
+	}
+	return views
+}
+
+func metricFormatter(m metricdata.Metrics) string {
+	return metric_format_prefix + m.Name
+}
+
+func gcpAttributeExpectedDefaults() []attribute.KeyValue {
+	return []attribute.KeyValue{
+		{Key: "location", Value: attribute.StringValue(mr_location_label_default)},
+		{Key: "cloud_platform", Value: attribute.StringValue(mr_cloud_platform_label_default)},
+		{Key: "host_id", Value: attribute.StringValue(mr_host_id_label_default)}}
+}
+
+func getPreparedResourceUsingGCPDetector(ctx context.Context, project string) (*resource.Resource, error) {
```

Comment:
nit: This func also seems superfluous and might be clearer in-line

Reply:
I wrote it this way to write tests; should I be doing something different to make it testable?